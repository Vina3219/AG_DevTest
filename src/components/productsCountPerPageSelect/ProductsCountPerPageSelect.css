.productsCountPerPageSelect {
  border: none;
  background-color: transparent;
  color: #8a8a8a;
  font-size: 16px;
  padding-right: 20px;
}
