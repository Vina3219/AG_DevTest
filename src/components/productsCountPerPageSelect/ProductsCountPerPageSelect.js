import React, { useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  selectProductsCountPerPage,
  setProductsCountPerPage,
} from "../../slice";
import { options } from "./constants";
import "./ProductsCountPerPageSelect.css";

const ProductsCountPerPageSelect = () => {
  const productsCountPerPage = useSelector(selectProductsCountPerPage);
  const dispatch = useDispatch();
  const handleSelectChange = useCallback(
    (event) => {
      const newProductsCountPerPage = Number(event.target.value);
      event.preventDefault();
      dispatch(
        setProductsCountPerPage({
          productsCountPerPage: newProductsCountPerPage,
        })
      );
    },
    [dispatch]
  );

  return (
    <select
      className="productsCountPerPageSelect"
      value={productsCountPerPage}
      onChange={handleSelectChange}
      data-testid="products-count-per-page-select"
    >
      {options.map((item) => (
        <option value={item.value} key={item.value}>
          {item.label}
        </option>
      ))}
    </select>
  );
};

export default ProductsCountPerPageSelect;
