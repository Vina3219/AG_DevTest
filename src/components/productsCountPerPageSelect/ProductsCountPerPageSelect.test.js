import { useSelector, useDispatch } from "react-redux";
import {
  selectProductsCountPerPage,
  setProductsCountPerPage,
} from "../../slice";
import { fireEvent, render, screen } from "@testing-library/react";
import ProductsCountPerPageSelect from "./ProductsCountPerPageSelect";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock("../../slice", () => ({
  ...jest.requireActual("../../slice"),
  selectProductsCountPerPage: jest.fn(),
  setProductsCountPerPage: jest.fn(),
}));

const mockedEvent = {
  target: {
    value: "4",
  },
  preventDefault: jest.fn(),
};

const dispatchSpy = jest.fn();

describe("ProductsCountPerPageSelect", () => {
  beforeEach(() => {
    useSelector.mockReset().mockImplementation((selector) => selector());
    selectProductsCountPerPage.mockReset().mockReturnValue(8);
    useDispatch.mockReset().mockReturnValue(dispatchSpy);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<ProductsCountPerPageSelect />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays the selected products count per page", () => {
    render(<ProductsCountPerPageSelect />);
    expect(screen.getByText("8 per page")).toBeInTheDocument();
  });

  describe("when selected option is changed", () => {
    it("displays the new value", () => {
      render(<ProductsCountPerPageSelect />);
      const select = screen.getByTestId("products-count-per-page-select");
      fireEvent.change(select, mockedEvent);
      expect(screen.getByText("4 per page")).toBeInTheDocument();
    });

    it("dispatches setProductsCountPerPage", () => {
      render(<ProductsCountPerPageSelect />);
      const select = screen.getByTestId("products-count-per-page-select");
      fireEvent.change(select, mockedEvent);
      expect(dispatchSpy).toBeCalledTimes(1);
      expect(setProductsCountPerPage).toBeCalledTimes(1);
      expect(setProductsCountPerPage).toBeCalledWith({
        productsCountPerPage: 4,
      });
    });
  });
});
