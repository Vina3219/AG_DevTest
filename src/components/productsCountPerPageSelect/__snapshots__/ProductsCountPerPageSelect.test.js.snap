// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsCountPerPageSelect matches snapshot 1`] = `
<DocumentFragment>
  <select
    class="productsCountPerPageSelect"
    data-testid="products-count-per-page-select"
  >
    <option
      value="4"
    >
      4 per page
    </option>
    <option
      value="6"
    >
      6 per page
    </option>
    <option
      value="8"
    >
      8 per page
    </option>
  </select>
</DocumentFragment>
`;
